Frameworks/CoreGraphics/DWriteFontBinaryDataLoader.mm

```diff
@@ -0,0 +1,119 @@
+
+    HRESULT RuntimeClassInitialize(CFDataRef data) {
+        CFRetain(data);
+        m_data.reset(data);
+
+        // Just use current time for m_lastWriteTime
+        FILETIME fileTime;
+        GetSystemTimeAsFileTime(&fileTime);
+
+        // Concat filetime into a single uint64_t
+        m_lastWriteTime = 0;
+        m_lastWriteTime |= static_cast<uint64_t>(fileTime.dwLowDateTime);
+        m_lastWriteTime |= static_cast<uint64_t>(fileTime.dwHighDateTime) << 32;
+        return S_OK;
+    }
+
+    HRESULT STDMETHODCALLTYPE GetFileSize(_Out_ uint64_t* fileSize) {
+        RETURN_HR_IF_NULL(E_INVALIDARG, fileSize);
+        *fileSize = CFDataGetLength(m_data.get());
+        return S_OK;
+    };
+
+    HRESULT STDMETHODCALLTYPE GetLastWriteTime(_Out_ uint64_t* lastWriteTime) {
+        RETURN_HR_IF_NULL(E_INVALIDARG, lastWriteTime);
+        *lastWriteTime = m_lastWriteTime;
+        return S_OK;
+    };
+
+    HRESULT STDMETHODCALLTYPE ReadFileFragment(_Out_ const void** fragmentStart,
+                                               uint64_t fileOffset,
+                                               uint64_t fragmentSize,
+                                               _Out_ void** fragmentContext) {
+        if (fileOffset + fragmentSize > CFDataGetLength(m_data.get())) {
+            return E_INVALIDARG;
+        }
+
+        if (fragmentStart) {
+            const uint8_t* underlyingBuffer = CFDataGetBytePtr(m_data.get());
+            *fragmentStart = reinterpret_cast<const void*>(&(underlyingBuffer[fileOffset]));
+        }
+
+        if (fragmentContext) {
+            // Deliberately unused: this is meant to be passed to ReleaseFileFragment() below to free part of the data,
+            // but this stream frees all the underlying CFData at once
+            *fragmentContext = nullptr;
+        }
+
+        return S_OK;
+    };
+
+    void STDMETHODCALLTYPE ReleaseFileFragment(void* fragmentContext){
+        // Deliberate no-op: data is released alongside the m_data member with the destruction of this object
+    };
+
+private:
+    woc::unique_cf<CFDataRef> m_data;
+    uint64_t m_lastWriteTime;
+};
+
+// DWriteFontBinaryDataLoader member functions
+
+DWriteFontBinaryDataLoader::DWriteFontBinaryDataLoader() {
+    // Deliberate no-op
+}
+
+HRESULT DWriteFontBinaryDataLoader::RuntimeClassInitialize(CFDataRef data) {
+    CFRetain(data);
+    m_data.reset(data);
+    return S_OK;
+}
+
+// Ignores first two params, just return the same kind of stream always
+// Stream returned is dictated by dataProvider passed in at initialization time
+HRESULT STDMETHODCALLTYPE DWriteFontBinaryDataLoader::CreateStreamFromKey(_In_ const void* fontFileReferenceKey,
+                                                                          uint32_t fontFileReferenceKeySize,
+                                                                          _Out_ IDWriteFontFileStream** fontFileStream) {
+    RETURN_HR_IF_NULL(E_INVALIDARG, fontFileStream);
```

Comment:
Usually __OUT__ invalid args are pretty much: E_POINTER
that's used  prominently  in windows.

---
In reply to: [90736872](https://github.com/Microsoft/WinObjC/pull/1504#discussion_r90736872) [](ancestors = 90736872)